atlasdb-commons/src/main/java/com/palantir/common/compression/GzipCompressingInputStream.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * (c) Copyright 2018 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.palantir.common.compression;
+
+import java.io.ByteArrayInputStream;
+import java.io.InputStream;
+import java.io.SequenceInputStream;
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.function.Supplier;
+import java.util.zip.CRC32;
+import java.util.zip.CheckedInputStream;
+import java.util.zip.Deflater;
+import java.util.zip.DeflaterInputStream;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Lists;
+import com.google.common.io.CountingInputStream;
+
+
+public final class GzipCompressingInputStream {
```

Comment:
we need to provide atlas with an input stream, and creating an inputstream out of an output stream is tricky (needs a thread to avoid locking up). PipedInputStream and PipedOutputStream help, but I don't want a thread.

Reply:
In terms of the documentation - I would be tempted to literally point out that just like, the person who thinks this is probably bogus and tries to reimplement will immediately figure out the issue :)